beluga/include/beluga/algorithm/estimation.hpp

```diff
@@ -35,6 +39,61 @@
 
 namespace beluga {
 
+namespace detail {
+
+/**
+ * \brief Given a range with quaternions and a normalized range of weights, compute the weighted average quaternion.
+ *
+ * \tparam QuaternionRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Eigen::Quaternion<Scalar>`.
+ * \tparam WeightsRange A [sized range](https://en.cppreference.com/w/cpp/ranges/sized_range) type whose
+ *  value type is `Scalar`.
+ * \tparam Scalar The scalar type, matching those of the weights.
+ * \param quaternions Range of quaternions to compute the weighted average from.
+ * \param normalized_weights Range of normalized weights. With matching indices wrt \c quaternions . Non-normalized
+ * weights will yield incorrect results. \return The average quaternion.
+ */
+template <
+    class QuaternionRange,
+    class WeightsRange,
+    class Scalar = typename ranges::range_value_t<QuaternionRange>::Scalar>
+Eigen::Quaternion<Scalar> weighted_average_quaternion(
+    const QuaternionRange& quaternions,
+    const WeightsRange& normalized_weights) {
+  // Implementation is based on https://github.com/strasdat/Sophus/blob/main/sophus/average.hpp#L133 with the variant of
+  // non-uniform weights.
```

Comment:
@serraramiro1 nit: consider adding a reference to the report / paper we know this is coming from, for future reference.

Reply:
Done